File: C#/Challenge/Portfolio.cs
namespace Challenge;

public class Portfolio
{
    /// <summary>
    /// Instance variable <c>_stocks</c> represents the stocks of the portfolio associated with the
    /// amount acquired.
    /// </summary>
    private readonly Dictionary<Stock, decimal> _stocks;

    /// <summary>
    /// Initialize a new Portfolio without stocks.
    /// </summary>
    public Portfolio()
    {
        _stocks = new Dictionary<Stock, decimal>();
    }

    /// <summary>
    /// Adds a certain <paramref name="amount"/> of the given <paramref name="stock"/> to the portfolio.
    /// </summary>
    /// <param name="stock">Stock to add.</param>
    /// <param name="amount">Amount to add of the given stock.</param>
    /// <returns>The current total amount of the stock in the portfolio.</returns>
    /// <exception cref="ArgumentOutOfRangeException">If the <paramref name="amount"/> is 0 or negative.</exception>
    public decimal AddStock(Stock stock, decimal amount)
    {
        if (amount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(amount), "The given amount must be above 0");
        }
        if (_stocks.ContainsKey(stock))
        {
            _stocks[stock] += amount;
        }
        else
        {
            _stocks.Add(stock, amount);
        }

        return _stocks[stock];
    }

    /// <summary>
    /// By default calculate the annualized return of the portfolio, if <paramref name="isAnnualized"/> is set to
    /// <c>false</c>, calculate the cumulative return instead.
    /// </summary>
    /// <param name="start">Initial date of the period.</param>
    /// <param name="end">Final date of the period.</param>
    /// <param name="isAnnualized">If <c>true</c> calculate the annualized return; otherwise, calculate the cumulative return.</param>
    /// <returns>The profit of the portfolio (annualized or cumulative).</returns>
    /// <exception cref="ArgumentException">If <paramref name="start"/> is a date after <paramref name="end"/>.</exception>
    public decimal Profit(DateOnly start, DateOnly end, bool isAnnualized = true)
    {
        if (end < start)
        {
            throw new ArgumentException($"The parameter {nameof(start)} is a date after {nameof(end)}");
        }
        
        return isAnnualized ?
            AnnualizedReturn(start, end) :
            CumulativeReturn(start, end);
    }

    /// <summary>
    /// Calculate the cumulative return between the dates given.<br/>
    /// See: <see href="https://www.investopedia.com/ask/answer/07/portfolio_calculations.asp#toc-calculating-the-percentage-return-of-your-portfolio">
    /// 'Calculating the Percentage Return of Your Portfolio' on www.investopedia.com</see><br/>
    /// See also: <seealso href="https://www.investopedia.com/terms/c/cumulativereturn.asp">
    /// 'Cumulative Return' on www.investopedia.com</seealso>
    /// </summary>
    /// <param name="start">Initial date of the period.</param>
    /// <param name="end">Final date of the period.</param>
    /// <returns>The cumulative return of the portfolio.</returns>
    private decimal CumulativeReturn(DateOnly start, DateOnly end)
    {
        if (start >= end || _stocks.Count <= 0)
        {
            return 0;
        }
        
        decimal startTotal = 0;
        decimal endTotal = 0;
        foreach (var stock in _stocks.Keys)
        {
            decimal stockAmount = _stocks[stock];
            startTotal += stockAmount * stock.Price(start);
            endTotal += stockAmount * stock.Price(end);
        }

        return (endTotal - startTotal) / startTotal;

    }

    /// <summary>
    /// Calculate the annualized return between the dates given.<br/>
    /// See: <see href="https://www.investopedia.com/terms/a/annualized-total-return.asp">'Annualized Total Return' on www.investopedia.com</see>
    /// </summary>
    /// <param name="start">Initial date of the period.</param>
    /// <param name="end">Final date of the period.</param>
    /// <returns>The annualized return of the portfolio.</returns>
    /// <exception cref="OverflowException">If the result of the calculation is over <c>decimal.MaxValue</c></exception>
    private decimal AnnualizedReturn(DateOnly start, DateOnly end)
    {
        decimal cumulativeReturn = CumulativeReturn(start, end);

        switch (cumulativeReturn)
        {
            // The difference of start and end dates is in years
            // Use the formula for yearly periods
            case > 0 when start.Month == end.AddDays(1).Month && start.Day == end.AddDays(1).Day:
            {
                decimal yearsHeld = end.AddDays(1).Year - start.Year;
                
                double basePow = (double) (1m + cumulativeReturn);
                double exponentPow = (double) (1m / yearsHeld);
                return (decimal)Math.Pow(basePow, exponentPow) - 1m;
            }
            // The difference of start and end dates isn't exact years
            // Use the formula with the difference of days instead
            case > 0:
            {
                DateTime dtStart = start.ToDateTime(TimeOnly.MinValue);
                DateTime dtEnd = end.ToDateTime(TimeOnly.MinValue);
                double daysHeld = (dtEnd - dtStart).Days;
                
                double basePow = (double) (1m + cumulativeReturn);
                double exponentPow = 365 / daysHeld;
                return (decimal)Math.Pow(basePow, exponentPow) - 1m;
            }
            default:
                return 0;
        }
    }

}

File: Python/portfolio/portfolio.py
# -*- coding: utf-8 -*-
from decimal import Decimal
from datetime import date, timedelta

from portfolio.stock import Stock


class Portfolio:
    """Portfolio which contains a dictionary with the stocks of a client.

    Attributes:
        _stocks (dict): Dictionary that correlates a stock with the amount that the client possesses (stock:amount).
    """

    def __init__(self):
        """Initialize a new Portfolio without stocks."""
        self._stocks = {}

    def add_stock(self, stock: Stock, amount: Decimal) -> Decimal:
        """Adds a certain ``amount`` of the given `stock`` to the portfolio.
        Args:
            stock: Stock to add.
            amount: Amount to add of the given stock.
        Returns:
            The current total amount of the stock in the portfolio.
        Raises:
            ValueError: If the ``amount`` is 0 or negative.
        """
        if amount <= 0:
            raise ValueError("The given amount must be above 0")
        if stock not in self._stocks:
            self._stocks[stock] = Decimal(0)
        self._stocks[stock] += amount
        return self._stocks[stock]

    def profit(self, start: date, end: date, is_annualized: bool = True) -> Decimal:
        """By default calculate the annualized return of the portfolio, if ``is_annualized``
        is set to False, calculate the cumulative return instead.
        Args:
            start: Initial date of the period.
            end: Final date of the period.
            is_annualized: If True calculate the annualized return; otherwise, calculate the cumulative return.
        Returns:
            The profit of the portfolio (annualized or cumulative).
        Raises:
            ValueError: If ``start`` is a date afeter ``end``.
        """
        if end < start:
            raise ValueError("The given end date is before start")
        if is_annualized:
            return self.annualized_return(start, end)
        return self.cumulative_return(start, end)

    def cumulative_return(self, start: date, end: date) -> Decimal:
        """Calculate the cumulative return between the dates given.

        See: `'Calculating the Percentage Return of Your Portfolio' on www.investopedia.com <https://www.investopedia.com/ask/answer/07/portfolio_calculations.asp#toc-calculating-the-percentage-return-of-your-portfolio>`_

        See also: `'Cumulative Return' on www.investopedia.com <https://www.investopedia.com/terms/c/cumulativereturn.asp>`_
        Args:
            start: Initial date of the period.
            end: Final date of the period.
        Returns:
            The cumulative return of the portfolio.
        """
        if start >= end or len(self._stocks) <= 0:
            return Decimal(0)
        else:
            start_total = Decimal(0)
            end_total = Decimal(0)
            for stock in self._stocks.keys():
                stock_amount = self._stocks[stock]
                start_total += (stock_amount * stock.price(start))
                end_total += (stock_amount * stock.price(end))
            return (end_total - start_total) / start_total

    def annualized_return(self, start: date, end: date) -> Decimal:
        """Calculate the annualized return between the dates given.

        See: `'Annualized Total Return' on www.investopedia.com <https://www.investopedia.com/terms/a/annualized-total-return.asp>`_
        Args:
            start: Initial date of the period.
            end: Final date of the period.
        Returns:
            The annualized return of the portfolio.
        """
        cumulative_return = self.cumulative_return(start, end)

        if cumulative_return <= 0:
            return Decimal(0)
        else:
            end_next_day = end + timedelta(days=1)
            # Period in years
            if start.month == end_next_day.month and start.day == end_next_day.day:
                years_held = Decimal(end_next_day.year - start.year)
                return (1 + cumulative_return) ** (Decimal(1) / years_held) - 1
            else:
                days_held = Decimal((end - start).days)
                return (1 + cumulative_return) ** (365 / days_held) - 1
